File: engine/notes.py
"""The thinking that is not the report.

Every report accumulates material that will never appear in it: the question to
put to the client, the paragraph that has to be rewritten once the pricing page
is archived, the reminder that the scorecard is still a guess. It has to live
somewhere, and the two bad answers are a separate app — where it is immediately
divorced from the report it belongs to — and the report itself, where it either
ships to the reader or gets deleted before it can.

So it lives beside the report, in two optional files:

    reports/<id>/notes.md    free prose. Anything.
    reports/<id>/todos.md    a markdown checklist.

Both are plain markdown on purpose. They are readable in any editor, diffable in
any review, greppable from any shell, and there is no schema to keep in sync with
anything — the same reasoning that makes folders the data model everywhere else
in this engine.

**Neither file is ever compiled into the PDF, and the citation rule does not
apply to either of them.** That is not an oversight, it is the point. The rule
exists so that nothing a *reader* sees can sit between a cited fact and a marked
opinion; a scratch pad has no reader but the author, and a half-formed thought
that had to be cited before it could be written down would simply not get
written down. `check` never opens these files, and `build` never sees them.

One thing does leak in the other direction, deliberately. A `// TODO:` or
`// FIXME:` left in a comment in `main.typ` is the same kind of note, written in
the place the thought occurred, and a list that omitted it would be a list you
could not trust to be complete — so they are harvested into the same view. They
are read-only here: `toggle` refuses on them, because a checkbox in a Typst
comment is prose, not state.
"""

from __future__ import annotations

import datetime as dt
import re
from collections.abc import Sequence
from dataclasses import dataclass, field
from pathlib import Path

from . import check
from .config import Config
from .workspace import Report, reports

TODOS_NAME = "todos.md"
NOTES_NAME = "notes.md"

# Where a task can have come from. Two of the three are writable; a marker in
# the report source is not, which `toggle` enforces.
SOURCES = (TODOS_NAME, NOTES_NAME, "main.typ")
WRITABLE = (TODOS_NAME, NOTES_NAME)

# A checklist item, as markdown actually gets written: any indentation (which is
# how nesting will arrive), any of the three bullet characters, and either mark.
# The indentation is captured and thrown away — this module reports a flat list,
# because a todo's parent is not information a checkbox click needs.
CHECKBOX = re.compile(r"^[ \t]*[-*+][ \t]+\[(?P<mark>[ xX])\][ \t]*(?P<text>.*)$")

# `// TODO:` and `// FIXME:` in the report source. Case-insensitive because a
# note written in a hurry is written in a hurry; the colon is required, so that
# a sentence merely containing the word does not become a task.
MARKER = re.compile(r"//+[ \t]*(?P<kind>TODO|FIXME)[ \t]*:[ \t]*(?P<text>[^\n]*)", re.I)

# `#tag` — a word, not a colour code and not a Typst directive. Only at a word
# boundary, so a URL fragment in a note is not a tag.
TAG = re.compile(r"(?:(?<=\s)|^)#(?P<tag>[A-Za-z][\w-]*)")

# `@2026-09-01` — the same sigil Typst uses for a citation, which is safe here
# precisely because these files are never compiled.
DUE = re.compile(r"(?:(?<=\s)|^)@(?P<date>\d{4}-\d{2}-\d{2})\b")

HEADING = """# Todo — {title}

Scratch, not evidence. This file is never compiled into the report and the
citation rule does not apply to it — that is what it is for.
"""

# What `report-maker new` seeds. The first item is the one piece of advice that
# changes how a report goes: the bibliography before the prose.
SEEDED = """
- [ ] Draft sources.yml before the prose.
- [ ] Decide what in this report is evidence and what is assessment.
"""


class NotesError(RuntimeError):
    """A refusal. The message names the fix, because a refusal that only says no
    teaches people to work around the tool."""


@dataclass
class Todo:
    """One task, wherever it was written."""

    text: str
    done: bool
    line: int
    tags: list[str] = field(default_factory=list)
    due: str | None = None
    source: str = TODOS_NAME  # "todos.md" | "main.typ" | "notes.md"


@dataclass
class Note:
    """`notes.md`, whole. There is nothing to parse — it is prose."""

    report: str
    path: Path
    text: str
    lines: int
    modified: str


# ── paths ────────────────────────────────────────────────────────────────────


def todos_file(report: Report) -> Path:
    return report.folder / TODOS_NAME


def notes_file(report: Report) -> Path:
    return report.folder / NOTES_NAME


def _read(path: Path) -> str | None:
    """The file's text, or None when it is not there.

    Decoding is lenient because these are scratch files: a stray byte pasted out
    of a PDF should cost one mangled character in a list, not an exception that
    takes down `scan` for every other report in the vault. `toggle` decodes
    strictly instead — it writes the result back.
    """
    if not path.is_file():
        return None
    return path.read_text(encoding="utf-8", errors="replace")


def _stamp(path: Path) -> str:
    return (
        dt.datetime.fromtimestamp(path.stat().st_mtime)
        .replace(microsecond=0)
        .isoformat(sep=" ")
    )


# ── parsing ──────────────────────────────────────────────────────────────────


def tags_in(text: str) -> list[str]:
    """Every `#tag`, in order, without repeats."""
    found: list[str] = []
    for match in TAG.finditer(text):
        tag = match.group("tag")
        if tag not in found:
            found.append(tag)
    return found


def due_in(text: str) -> str | None:
    """The first `@YYYY-MM-DD` that is a real date.

    A token that looks like a date but is not one (`@2026-13-40`) yields None
    rather than an error: it is a typo in a scratch file, and refusing to list
    the item would be a strange punishment for it.
    """
    for match in DUE.finditer(text):
        try:
            return dt.date.fromisoformat(match.group("date")).isoformat()
        except ValueError:
            continue
    return None


def _item(text: str, done: bool, line: int, source: str) -> Todo:
    text = text.rstrip()
    return Todo(
        text=text,
        done=done,
        line=line,
        tags=tags_in(text),
        due=due_in(text),
        source=source,
    )


def parse_checklist(text: str, source: str = TODOS_NAME) -> list[Todo]:
    """Every `- [ ]` / `- [x]` line in a markdown file."""
    found: list[Todo] = []
    for number, line in enumerate(text.split("\n"), 1):
        match = CHECKBOX.match(line)
        if match:
            found.append(
                _item(
                    match.group("text"),
                    done=match.group("mark").lower() == "x",
                    line=number,
                    source=source,
                )
            )
    return found


def harvest(raw: str) -> list[Todo]:
    """`// TODO:` and `// FIXME:` comments in Typst source.

    Whether a `//` opens a comment is a question the linter already answers, and
    answering it twice is how two parsers start disagreeing about what a report
    says. So this reuses `check.scrub`, which blanks comments to spaces while
    keeping every offset: a marker is in a comment exactly when the character
    under it came back blank.

    `scrub` also blanks code blocks, so a `// TODO:` inside one is harvested
    too. That false positive costs one line in a list; a second comment parser
    would cost a permanent divergence from the rule the linter enforces.
    """
    blanked = check.scrub(raw)
    found: list[Todo] = []
    for match in MARKER.finditer(raw):
        if blanked[match.start()] != " ":
            continue  # a `//` in prose, or the one in `https://`
        # The marker word goes into the text: TODO and FIXME do not mean the
        # same thing, and the dataclass has nowhere else to keep the difference.
        kind = match.group("kind").upper()
        rest = match.group("text").strip()
        text = f"{kind}: {rest}" if rest else kind
        found.append(
            _item(text, done=False, line=check.line_of(raw, match.start()), source="main.typ")
        )
    return found


# ── reading one report ───────────────────────────────────────────────────────


def todos(report: Report) -> list[Todo]:
    """Every task attached to a report, in the order a person would look for
    them: the checklist first, then anything in the notes, then the report
    source itself."""
    found: list[Todo] = []
    text = _read(todos_file(report))
    if text is not None:
        found += parse_checklist(text, TODOS_NAME)
    text = _read(notes_file(report))
    if text is not None:
        found += parse_checklist(text, NOTES_NAME)
    if report.main.is_file():
        found += harvest(report.main.read_text(encoding="utf-8", errors="replace"))
    return found


def notes(report: Report) -> Note | None:
    """`notes.md`, or None. An absent scratch pad is not a finding."""
    path = notes_file(report)
    text = _read(path)
    if text is None:
        return None
    return Note(
        report=report.id,
        path=path,
        text=text,
        lines=len(text.splitlines()),
        modified=_stamp(path),
    )


def _modified(report: Report) -> str | None:
    """The most recent of the two files, so the app can sort by "last touched"
    without opening either."""
    stamps = [
        _stamp(path)
        for path in (todos_file(report), notes_file(report))
        if path.is_file()
    ]
    return max(stamps) if stamps else None


# ── reading a vault ──────────────────────────────────────────────────────────


def one(cfg: Config, target: str) -> Report:
    """The single report a write is aimed at.

    Adding a task to eight reports because a folder name matched is not a thing
    anybody meant, so anything that writes goes through here first.
    """
    found = reports(cfg, target)
    if not found:
        raise NotesError(
            'this vault has no reports yet — `report-maker new "Title"` makes one'
        )
    if len(found) != 1:
        raise NotesError(
            f"{target!r} matches {len(found)} reports — name exactly one: "
            + ", ".join(report.id for report in found)
        )
    return found[0]


def scan(cfg: Config, target: str | None = None, *, open_only: bool = False) -> list[dict]:
    """One row per report that has something on its pad.

    Reports with neither file and no harvested marker are left out entirely: a
    list of eighty reports, seventy-six of them empty, is a list nobody reads.

    `open` and `done` always count everything found, even under `open_only` —
    the counts describe the report, and the list describes the filter. Reporting
    `done: 0` for a report with nine finished tasks would be a lie told to make
    a flag look tidy.
    """
    rows: list[dict] = []
    for report in reports(cfg, target):
        found = todos(report)
        note = notes(report)
        shown = [todo for todo in found if not todo.done] if open_only else found
        if not shown and note is None:
            continue
        rows.append(
            {
                "id": report.id,
                "open": sum(1 for todo in found if not todo.done),
                "done": sum(1 for todo in found if todo.done),
                "todos": to_json(shown),
                "has_notes": note is not None,
                "modified": _modified(report),
            }
        )
    return rows


# ── writing ──────────────────────────────────────────────────────────────────


def toggle(report: Report, line: int, done: bool, *, source: str = TODOS_NAME) -> None:
    """Flip one checkbox, rewriting exactly one character on exactly one line.

    Surgical on purpose. This is called from a checkbox click in the app while
    the same file may be open in an editor, and a whole-file rewrite would eat
    whatever the person typed a second earlier. Everything outside the two
    characters between the brackets comes back byte for byte — line endings,
    trailing whitespace, the lot — and when the box is already in the requested
    state nothing is written at all, so a double click does not bump the mtime a
    watcher is keyed on.
    """
    if source == "main.typ":
        raise NotesError(
            "a `// TODO:` in main.typ is prose, not state — it has no checkbox to "
            "flip.\n"
            "  Edit the line in the report, or move the task to "
            f"{TODOS_NAME} where it can be ticked."
        )
    if source not in WRITABLE:
        raise NotesError(f"a tickable task lives in {' or '.join(WRITABLE)}, not in {source!r}")

    path = report.folder / source
    # Messages name the file the way the vault does — an absolute path from a
    # subprocess is noise in an app's error toast.
    where = check.relative(path, report.cfg.root)
    if not path.is_file():
        raise NotesError(f"{where} does not exist — add a task first")

    try:
        text = path.read_bytes().decode("utf-8")
    except UnicodeDecodeError as exc:
        # Lenient decoding is fine for reading; writing back a replaced
        # character would silently destroy whatever was really there.
        raise NotesError(f"{where} is not valid UTF-8 ({exc.reason}) — fix it by hand") from exc

    # Split on "\n" rather than with splitlines(), which also breaks on \x0b,
    # \x0c and U+2028, and would therefore number lines differently from every
    # other tool that has ever reported a line in this file.
    lines = text.split("\n")
    if not 1 <= line <= len(lines):
        raise NotesError(f"{where} has {len(lines)} lines — there is no line {line}")

    original = lines[line - 1]
    match = CHECKBOX.match(original)
    if not match:
        raise NotesError(
            f"{where}:{line} is not a checklist item: {original.strip()!r}\n"
            "  A task is a line like `- [ ] the thing to do`."
        )

    start, end = match.span("mark")
    mark = "x" if done else " "
    if original[start:end] == mark:
        return
    lines[line - 1] = original[:start] + mark + original[end:]
    path.write_bytes("\n".join(lines).encode("utf-8"))


def add(report: Report, text: str) -> Todo:
    """Append a task, creating `todos.md` with its heading if it is not there."""
    text = text.strip()
    if not text:
        raise NotesError("a task needs some words")
    if "\n" in text:
        raise NotesError(
            "a task is one line — split it into two tasks, or put the paragraph in "
            f"{NOTES_NAME}"
        )
    # Somebody pasting a line out of a markdown file will bring the bullet with
    # it; appending it verbatim would produce `- [ ] - [ ] …`.
    text = re.sub(r"^[-*+][ \t]+(\[[ xX]\][ \t]*)?", "", text)

    path = todos_file(report)
    existing = _read(path)
    if existing is None:
        existing = HEADING.format(title=report.id) + "\n"
    elif existing and not existing.endswith("\n"):
        existing += "\n"

    line = existing.count("\n") + 1
    path.write_text(existing + f"- [ ] {text}\n", encoding="utf-8")
    return _item(text, done=False, line=line, source=TODOS_NAME)


def starter_text(report_id: str) -> str:
    """What `report-maker new` writes into a fresh report folder."""
    return HEADING.format(title=report_id) + SEEDED


# ── output ───────────────────────────────────────────────────────────────────


def to_json(items: Sequence[Todo]) -> list[dict]:
    return [
        {
            "text": todo.text,
            "done": todo.done,
            "line": todo.line,
            "tags": todo.tags,
            "due": todo.due,
            "source": todo.source,
        }
        for todo in items
    ]


def note_json(note: Note | None, *, root: Path | None = None) -> dict | None:
    if note is None:
        return None
    return {
        "report": note.report,
        "path": check.relative(note.path, root) if root else note.path.as_posix(),
        "text": note.text,
        "lines": note.lines,
        "modified": note.modified,
    }


def _flag(todo: dict, today: dt.date) -> str:
    """Overdue and due-today, said out loud. A due date nobody surfaces is a
    comment, and this is the whole of what the date is for."""
    if todo["done"] or not todo["due"]:
        return ""
    due = dt.date.fromisoformat(todo["due"])
    if due < today:
        return "overdue"
    if due == today:
        return "due today"
    return ""


def report_todos(cfg: Config, rows: Sequence[dict]) -> int:
    """Print the pad, grouped by report.

    Always exits 0. An open task is not a broken build — it is the note that
    says the build is not finished yet, which is a different thing, and a
    command that failed on one would train people to stop writing them down.
    """
    if not rows:
        print(
            "  nothing on the pad\n"
            f"  `report-maker todos <report> --add \"…\"` starts reports/<id>/{TODOS_NAME}"
        )
        return 0

    today = dt.date.today()
    labels = [
        f"[{'x' if todo['done'] else ' '}] {todo['text']}"
        for row in rows
        for todo in row["todos"]
    ]
    width = min(max((len(label) for label in labels), default=0), 64)

    for row in rows:
        print(f"  {(cfg.reports / row['id']).relative_to(cfg.root)}")
        for todo in row["todos"]:
            label = f"[{'x' if todo['done'] else ' '}] {todo['text']}"
            where = f"{todo['source']}:{todo['line']}"
            print(f"    {label:<{width}}  {where:<16}  {_flag(todo, today)}".rstrip())
        if row["has_notes"]:
            print(f"    {NOTES_NAME}  (last touched {row['modified']})")

    total_open = sum(row["open"] for row in rows)
    total_done = sum(row["done"] for row in rows)
    noted = sum(1 for row in rows if row["has_notes"])
    summary = f"\n  {total_open} open, {total_done} done across {len(rows)} report(s)"
    if noted:
        summary += f" · {noted} with notes"
    print(summary)
    return 0

File: app/src/renderer/src/lib/notes.ts
/**
 * The pad beside a report: `todos.md` and `notes.md`.
 *
 * Two files, both optional, both plain markdown, neither ever compiled into the
 * PDF — see `engine/notes.py` for why they live beside the report rather than in
 * a separate app. The citation rule does not reach them, which is the point: a
 * half-formed thought that had to be cited before it could be written down would
 * simply not get written down.
 *
 * The split of responsibility here is deliberate and is the only interesting
 * thing in this file.
 *
 * - **The list is the engine's.** `todos --json` decides what a task is, which
 *   `#tag` and `@date` it carries, and which `// TODO:` in `main.typ` counts as a
 *   comment rather than prose (that last one reuses the linter's own scrubber, so
 *   the pad and `check` can never disagree about what a comment is). Nothing here
 *   parses a checklist.
 * - **Ticking a box is the engine's.** `toggle` rewrites one character on one
 *   line and leaves the rest of the file byte for byte, which matters because the
 *   same file may be open in the editor while a checkbox is clicked. A whole-file
 *   write from the app would eat whatever was typed a second earlier.
 * - **Editing the notes is not.** `notes.md` is prose in a text file and the app
 *   already has a guarded reader and writer for those. Routing a textarea through
 *   a subprocess would add a spawn per autosave and answer no question the app
 *   could not answer itself; there is no rule about a vault in reading a file.
 */

import { useCallback, useEffect, useMemo, useRef, useState } from 'react'
import { describeError } from '@/lib/sources'

// ── What the engine reports ──────────────────────────────────────────────────
//
// These mirror `engine/notes.py` exactly — `Todo`, and the row `scan()` builds
// per report. Declared here rather than in shared/types.ts only because that file
// has an owner; they belong beside the rest of the engine's shapes once the two
// land together.

/** Which file a task was written in. `main.typ` items are harvested comments. */
export type TodoSource = 'todos.md' | 'notes.md' | 'main.typ'

export type Todo = {
  text: string
  done: boolean
  /** 1-based line in `source`, so a click can jump to it. */
  line: number
  tags: string[]
  /** `YYYY-MM-DD`, or null. The engine drops a date that is not a real one. */
  due: string | null
  source: TodoSource
}

/** One report with something on its pad. Reports with neither file are absent. */
export type TodoReport = {
  id: string
  /** Open and done count everything found, whatever the list happens to show. */
  open: number
  done: number
  todos: Todo[]
  has_notes: boolean
  /** The later of the two files' mtimes, `YYYY-MM-DD HH:MM:SS`, or null. */
  modified: string | null
}

export type TodosResponse = { reports: TodoReport[] }

// ── Where the pad lives ──────────────────────────────────────────────────────
//
// Vault-relative POSIX, the form `Finding.path`, the sources panel and the search
// results all speak, so a jump from this panel takes the same shape as any other.

export function todosPath(reportId: string): string {
  return `reports/${reportId}/todos.md`
}

export function notesPath(reportId: string): string {
  return `reports/${reportId}/notes.md`
}

/** The file a task was written in — `todos.md`, `notes.md` or the report itself. */
export function pathOf(reportId: string, todo: Pick<Todo, 'source'>): string {
  return `reports/${reportId}/${todo.source}`
}

// ── The commands ─────────────────────────────────────────────────────────────
//
// Written down once, here, so the panel never assembles an argv inline and the
// shape the app depends on is readable in one place.

export function listArgs(target?: string | null): string[] {
  return ['todos', ...(target ? [target] : []), '--json']
}

export function addArgs(reportId: string, text: string): string[] {
  return ['todos', reportId, '--add', text]
}

/**
 * Tick or untick one box.
 *
 * `--in` names the file because a checklist item may sit in `notes.md` as easily
 * as in `todos.md`, and the engine needs to know which one it is rewriting. A
 * `main.typ` task never reaches here: `notes.toggle` refuses it, and the panel
 * does not offer a checkbox for one.
 */
export function toggleArgs(
  reportId: string,
  line: number,
  done: boolean,
  source: TodoSource = 'todos.md'
): string[] {
  return ['todos', reportId, done ? '--check' : '--uncheck', String(line), '--in', source]
}

// ── Reading ──────────────────────────────────────────────────────────────────

/**
 * The rows out of `todos --json`.
 *
 * Accepts the wrapped form and a bare list. The wrapper is the CLI's — `scan()`
 * returns the list itself — and a reader that only understood one of them would
 * break on whichever one it was not written against.
 */
export function readRows(response: TodosResponse | TodoReport[] | null): TodoReport[] {
  const rows = Array.isArray(response) ? response : (response?.reports ?? [])
  return Array.isArray(rows) ? rows : []
}

function loadRows(vault: string, target?: string | null): Promise<TodoReport[]> {
  return window.api.engine
    .json<TodosResponse | TodoReport[]>(vault, listArgs(target))
    .then(readRows)
}

/**
 * `notes.md`, as text.
 *
 * An absent file and an empty one read the same in a textarea but not in the
 * panel: one is a scratch pad nobody has started, the other is one somebody
 * emptied. Both come back, so the caller can tell them apart.
 */
async function loadNotes(
  vault: string,
  reportId: string
): Promise<{ text: string; exists: boolean }> {
  const path = `${vault}/${notesPath(reportId)}`
  if (!(await window.api.files.exists(vault, path))) return { text: '', exists: false }
  return { text: await window.api.files.read(vault, path), exists: true }
}

// ── Dates ────────────────────────────────────────────────────────────────────

/** Today, in the local zone. `Date#toISOString` is UTC and would call a task
 *  overdue an evening early for anyone west of Greenwich. */
export function localToday(): string {
  const now = new Date()
  const month = String(now.getMonth() + 1).padStart(2, '0')
  const day = String(now.getDate()).padStart(2, '0')
  return `${now.getFullYear()}-${month}-${day}`
}

export type DueState = 'overdue' | 'today' | null

/**
 * How a due date reads today — the app's half of `engine/notes.py:_flag`, which
 * says the same thing on the terminal. A done task has no due state: it was
 * finished, and telling somebody they are late for something they have already
 * done is the fastest way to get the dates ignored.
 *
 * ISO dates compare correctly as strings, which is the whole reason the engine
 * emits them in that form.
 */
export function dueState(todo: Todo, today: string = localToday()): DueState {
  if (todo.done || !todo.due) return null
  if (todo.due < today) return 'overdue'
  if (todo.due === today) return 'today'
  return null
}

/** `2026-08-18 09:12:33` → `09:12` when that was today, else `2026-08-18`. The
 *  seconds are never the thing being read. */
export function touched(modified: string | null | undefined): string {
  if (!modified) return ''
  const [day, time = ''] = modified.split(' ')
  return day === localToday() ? time.slice(0, 5) : day
}

// ── Text ─────────────────────────────────────────────────────────────────────

/**
 * The task without the tokens that are about to be drawn as chips.
 *
 * Only the exact `#tag`s and the one `@date` the engine reported are removed, so
 * this is a presentation of the engine's answer rather than a second parse of the
 * line: a `#` the engine did not call a tag stays in the text where the writer
 * put it. Leaving them in would say everything twice; stripping by our own rule
 * would eventually strip something the engine kept.
 */
export function label(todo: Todo): string {
  let text = todo.text
  for (const tag of todo.tags) {
    // Tags are `[A-Za-z][\w-]*` — nothing in one is a regex metacharacter.
    text = text.replace(new RegExp(`(^|\\s)#${tag}(?![\\w-])`), '$1')
  }
  if (todo.due) text = text.replace(`@${todo.due}`, '')
  return text.replace(/\s{2,}/g, ' ').trim()
}

// ── Counting ─────────────────────────────────────────────────────────────────

export function openCount(todos: readonly Todo[]): number {
  return todos.reduce((count, todo) => count + (todo.done ? 0 : 1), 0)
}

export function overdueCount(todos: readonly Todo[], today: string = localToday()): number {
  return todos.reduce((count, todo) => count + (dueState(todo, today) === 'overdue' ? 1 : 0), 0)
}

/** One optimistic tick, counts and all. Matching on line *and* file because the
 *  same line number exists in `todos.md` and in `notes.md`. */
function flip(row: TodoReport | null, line: number, source: TodoSource, done: boolean): TodoReport | null {
  if (!row) return row
  let moved = false
  const todos = row.todos.map((todo) => {
    if (todo.line !== line || todo.source !== source || todo.done === done) return todo
    moved = true
    return { ...todo, done }
  })
  if (!moved) return row
  const step = done ? 1 : -1
  return { ...row, todos, done: row.done + step, open: row.open - step }
}

// ── One report's pad ─────────────────────────────────────────────────────────

export type UseNotes = {
  /** Every task attached to the report, in the engine's order: the checklist,
   *  then anything in the notes, then the report source. */
  todos: Todo[]
  open: number
  done: number
  /** `notes.md` as it is on disk; empty when there is no file. */
  notes: string
  /** Whether the file exists at all, which an empty string cannot say. */
  hasNotes: boolean
  modified: string | null
  loading: boolean
  /** A write is in flight — a tick, or a new task. */
  busy: boolean
  /** Whatever last went wrong, in the engine's own words. */
  error: string | null
  reload: () => void
  toggle: (line: number, done: boolean, source?: TodoSource) => Promise<void>
  /** Resolves true when the task landed. The field keeps what was typed when it
   *  did not — a refusal that costs you the sentence teaches you not to write
   *  the next one down. */
  add: (text: string) => Promise<boolean>
  /** Write `notes.md`. Throws on failure so the editor can say it did not save. */
  saveNotes: (text: string) => Promise<void>
}

const EMPTY: Todo[] = []

/**
 * The pad for one report, kept current.
 *
 * `revision` is the parent's "something happened" counter — a build, a save, a
 * vault switch — exactly as `useSources` uses it. The hook is meant to be owned
 * by the parent rather than by the panel, because the status-bar chip needs the
 * same count and the panel is a sidebar tab that only exists while it is
 * selected: a chip fed by a hook inside a hidden tab would show nothing.
 */
export function useNotes(
  vault: string | null,
  reportId: string | null,
  revision = 0
): UseNotes {
  const [row, setRow] = useState<TodoReport | null>(null)
  const [notes, setNotes] = useState('')
  const [hasNotes, setHasNotes] = useState(false)
  const [loading, setLoading] = useState(false)
  const [busy, setBusy] = useState(false)
  const [error, setError] = useState<string | null>(null)
  const [nonce, setNonce] = useState(0)

  const reload = useCallback(() => setNonce((n) => n + 1), [])

  // A promise cannot be cancelled, only disowned: switching report while a slow
  // answer is in flight must not paint the previous report's pad over the new one.
  const generation = useRef(0)

  useEffect(() => {
    const run = ++generation.current

    if (!vault || !reportId) {
      setRow(null)
      setNotes('')
      setHasNotes(false)
      setError(null)
      setLoading(false)
      return
    }

    setLoading(true)
    void (async () => {
      try {
        const [rows, scratch] = await Promise.all([
          loadRows(vault, reportId),
          loadNotes(vault, reportId)
        ])
        if (run !== generation.current) return
        setRow(rows.find((entry) => entry.id === reportId) ?? null)
        setNotes(scratch.text)
        setHasNotes(scratch.exists)
        setError(null)
      } catch (err) {
        if (run !== generation.current) return
        setRow(null)
        setNotes('')
        setHasNotes(false)
        setError(describeError(err))
      } finally {
        if (run === generation.current) setLoading(false)
      }
    })()

    return () => {
      generation.current += 1
    }
  }, [vault, reportId, revision, nonce])

  // `saveNotes` has to stay stable per report so the editor can flush its last
  // keystrokes on the way out through the binding they were typed under; reading
  // existence from a ref keeps it out of the dependency list.
  const exists = useRef(hasNotes)
  exists.current = hasNotes

  /**
   * Tick or untick, painted before the engine is asked.
   *
   * A checkbox that waits for a subprocess feels broken, so the flip happens
   * locally and is undone if the engine refuses — and it does refuse, on a
   * `main.typ` comment, on a line that is not a checklist item, on a file that is
   * not valid UTF-8. The rollback flips back rather than restoring a snapshot,
   * which is the form that survives two clicks landing at once.
   */
  const toggle = useCallback(
    async (line: number, done: boolean, source: TodoSource = 'todos.md') => {
      if (!vault || !reportId) return
      setRow((current) => flip(current, line, source, done))
      setBusy(true)
      const result = await window.api.engine.run(vault, toggleArgs(reportId, line, done, source))
      setBusy(false)
      if (result.code !== 0) {
        setRow((current) => flip(current, line, source, !done))
        setError((result.stderr || result.stdout || `exit ${result.code}`).trimEnd())
        return
      }
      setError(null)
      reload()
    },
    [vault, reportId, reload]
  )

  /**
   * Append a task.
   *
   * Not optimistic, unlike the tick: the engine chooses the line it lands on,
   * creates `todos.md` with its heading when there is none, and strips a bullet
   * off a line pasted out of another markdown file. A row invented here would
   * carry a line number that does not exist yet, and the first click on its
   * checkbox would rewrite the wrong line.
   */
  const add = useCallback(
    async (text: string): Promise<boolean> => {
      const task = text.trim()
      if (!vault || !reportId || !task) return false
      setBusy(true)
      const result = await window.api.engine.run(vault, addArgs(reportId, task))
      setBusy(false)
      if (result.code !== 0) {
        setError((result.stderr || result.stdout || `exit ${result.code}`).trimEnd())
        return false
      }
      setError(null)
      reload()
      return true
    },
    [vault, reportId, reload]
  )

  const saveNotes = useCallback(
    async (text: string) => {
      if (!vault || !reportId) return
      // Focusing an empty textarea is not a reason to create a file. Once the
      // file exists, emptying it is a real edit and is written.
      if (!exists.current && text.trim() === '') return
      try {
        await window.api.files.write(vault, `${vault}/${notesPath(reportId)}`, text)
      } catch (err) {
        setError(describeError(err))
        throw err
      }
      // The hook's own copy moves with the write, so the next reload agrees with
      // what is on screen instead of racing the editor back to an older text.
      setNotes(text)
      setHasNotes(true)
    },
    [vault, reportId]
  )

  return {
    todos: row?.todos ?? EMPTY,
    open: row?.open ?? 0,
    done: row?.done ?? 0,
    notes,
    hasNotes,
    modified: row?.modified ?? null,
    loading,
    busy,
    error,
    reload,
    toggle,
    add,
    saveNotes
  }
}

// ── The whole vault ──────────────────────────────────────────────────────────

export type UseVaultTodos = {
  /** Only reports with something on the pad; the engine leaves the rest out. */
  reports: TodoReport[]
  /** Every task in the vault, flattened — what a chip counts. */
  todos: Todo[]
  /** Report id → its row, for a dashboard card that wants one number. */
  byId: Map<string, TodoReport>
  open: number
  overdue: number
  loading: boolean
  error: string | null
  reload: () => void
}

const NO_ROWS: TodoReport[] = []

/**
 * The pad across the vault, in one subprocess.
 *
 * `todos --json` with no target already walks every report, so the dashboard
 * badges every card from a single call rather than one per card. The reports it
 * omits are the ones with nothing on the pad, and `byId` returning undefined for
 * those is the answer, not a gap.
 */
export function useVaultTodos(vault: string | null, revision = 0): UseVaultTodos {
  const [reports, setReports] = useState<TodoReport[]>(NO_ROWS)
  const [loading, setLoading] = useState(false)
  const [error, setError] = useState<string | null>(null)
  const [nonce, setNonce] = useState(0)

  const reload = useCallback(() => setNonce((n) => n + 1), [])

  useEffect(() => {
    if (!vault) {
      setReports(NO_ROWS)
      setError(null)
      setLoading(false)
      return
    }
    let stale = false
    setLoading(true)
    loadRows(vault)
      .then((rows) => {
        if (stale) return
        setReports(rows)
        setError(null)
      })
      .catch((err) => {
        if (stale) return
        setReports(NO_ROWS)
        setError(describeError(err))
      })
      .finally(() => {
        if (!stale) setLoading(false)
      })
    return () => {
      stale = true
    }
  }, [vault, revision, nonce])

  const byId = useMemo(
    () => new Map(reports.map((row) => [row.id, row] as const)),
    [reports]
  )
  const todos = useMemo(() => reports.flatMap((row) => row.todos), [reports])
  // Summed from the rows rather than from `todos`, because a row counts every
  // task it found while the list it carries may have been filtered by the engine.
  const open = useMemo(() => reports.reduce((count, row) => count + row.open, 0), [reports])
  const overdue = useMemo(() => overdueCount(todos), [todos])

  return { reports, todos, byId, open, overdue, loading, error, reload }
}
